### PR TITLE
Implement item hooks for potions and enchantments

### PR DESCRIPTION
**Pull request by Random832**
Add hooks for enchantments and potions to determine what tabs they go in for creative mode (or whether to appear in creative mode at all), and for potions to control whether the potion bottle has the enchanted 'foil' appearance.

### CODE REVIEW

**Comment by BosseYT** [comment]
Its cool

**Comment by Shadows-of-Fire** [comment]
I think I would prefer `allowedInCreativeTab` be renamed to `isAllowedInCreativeTab`, `isAllowedInTab`, or `isAllowedIn`, but otherwise looks good to me.

File: src/main/java/net/minecraftforge/common/extensions/IForgeEnchantment.java [hunk 1]
```diff
@@ -28,4 +28,19 @@ default float getDamageBonus(int level, MobType mobType, ItemStack enchantedItem
     {
         return self().getDamageBonus(level, mobType);
     }
+
+    /**
+     * Should the book for this enchantment appear in the creative tab
+     * @param book Items.ENCHANTED_BOOK normally, but modded items may also use this hook.
+     * @param tab The creative tab that items are being added to
+     * @return true to put the item variant in the tab.
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/extensions/IForgeEnchantment.java]:
```suggestion
     * @return whether the given Item's variant for this potion should appear in the respective creative tab
```

File: src/main/java/net/minecraftforge/common/extensions/IForgeEnchantment.java [hunk 2]
```diff
@@ -28,4 +28,19 @@ default float getDamageBonus(int level, MobType mobType, ItemStack enchantedItem
...
+     */
+    default boolean allowedInCreativeTab(net.minecraft.world.item.Item book, net.minecraft.world.item.CreativeModeTab tab) {
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/extensions/IForgeEnchantment.java]: Please use imports instead of FQNs.
Formatting: Braces on new lines.

File: src/main/java/net/minecraftforge/common/extensions/IForgeEnchantment.java [hunk 3]
```diff
@@ -28,4 +28,19 @@ default float getDamageBonus(int level, MobType mobType, ItemStack enchantedItem
...
+        if(!self().isAllowedOnBooks())
+            return false;
+        else if(tab == net.minecraft.world.item.CreativeModeTab.TAB_SEARCH)
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/extensions/IForgeEnchantment.java]: Please use imports instead of FQNs.

File: src/main/java/net/minecraftforge/common/extensions/IForgeEnchantment.java [hunk 4]
```diff
@@ -28,4 +28,19 @@ default float getDamageBonus(int level, MobType mobType, ItemStack enchantedItem
...
+            return self().category != null;
+        else
+            return tab.hasEnchantmentCategory(self().category);
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/extensions/IForgeEnchantment.java]: Please do not use if statements without braces.
Please ensure the correct formatting (space missing after `if`)

File: src/main/java/net/minecraftforge/common/extensions/IForgePotion.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
+package net.minecraftforge.common.extensions;
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/extensions/IForgePotion.java:6]: Missing license header.

File: src/main/java/net/minecraftforge/common/extensions/IForgePotion.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import net.minecraft.world.item.alchemy.Potion;
+import net.minecraft.world.item.alchemy.PotionUtils;
+
+public interface IForgePotion {
+    private Potion self() {
+        return (Potion)this;
+    }
+
+    /**
+     * Determines whether the potion-related item [bottle, tipped arrow, or a modded item] should appear in a creative tab.
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/extensions/IForgePotion.java]: Why brackets instead of normal parentheses?

File: src/main/java/net/minecraftforge/common/extensions/IForgePotion.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
+package net.minecraftforge.common.extensions;
+
+import net.minecraft.world.item.CreativeModeTab;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.alchemy.Potion;
+import net.minecraft.world.item.alchemy.PotionUtils;
+
+public interface IForgePotion {
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/extensions/IForgePotion.java]: For the whole class: braces on new lines.

File: src/main/java/net/minecraftforge/common/extensions/IForgePotion.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    private Potion self() {
+        return (Potion)this;
+    }
+
+    /**
+     * Determines whether the potion-related item [bottle, tipped arrow, or a modded item] should appear in a creative tab.
+     * 
+     * @param item The item being added to a creative tab
+     * @param tab The tab items are being added to
+     * @param isDefaultTab The tab is the default tab for this item [e.g. Brewing for bottles, Combat for tipped arrows]
```
Comment (diesieben07) [inline src/main/java/net/minecraftforge/common/extensions/IForgePotion.java]:
```suggestion
     * @param isDefaultTab whether the tab is the default tab for this item (e.g. Brewing for bottles, Combat for tipped arrows)
```

File: patches/minecraft/net/minecraft/world/item/TippedArrowItem.java.patch [hunk 9]
```diff
@@ -0,0 +1,13 @@
+--- a/net/minecraft/world/item/TippedArrowItem.java
++++ b/net/minecraft/world/item/TippedArrowItem.java
+@@ -20,8 +_,8 @@
+    }
+ 
+    public void m_6787_(CreativeModeTab p_43356_, NonNullList<ItemStack> p_43357_) {
+-      if (this.m_220152_(p_43356_)) {
+-         for(Potion potion : Registry.f_122828_) {
++      for (Potion potion : Registry.f_122828_) {
++         if (potion.allowedInCreativeTab(this, p_43356_, this.m_220152_(p_43356_))) {
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/world/item/TippedArrowItem.java.patch:10]: Why did you remove the `allowedIn` call? This should be a one-line addition patch that simply patches in the if statement.

File: patches/minecraft/net/minecraft/world/item/PotionItem.java.patch [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    public boolean m_5812_(ItemStack p_42999_) {
+-      return super.m_5812_(p_42999_) || !PotionUtils.m_43547_(p_42999_).isEmpty();
++      return super.m_5812_(p_42999_) || PotionUtils.m_43579_(p_42999_).isFoil(p_42999_);
+    }
+ 
+    public void m_6787_(CreativeModeTab p_42981_, NonNullList<ItemStack> p_42982_) {
+-      if (this.m_220152_(p_42981_)) {
+-         for(Potion potion : Registry.f_122828_) {
++      for(Potion potion : Registry.f_122828_) {
++         if (potion.allowedInCreativeTab(this, p_42981_, this.m_220152_(p_42981_))) {
```
Comment (diesieben07) [inline patches/minecraft/net/minecraft/world/item/PotionItem.java.patch:15]: Why did you remove the `allowedIn` call? This should be a one-line addition patch that simply patches in the if statement.

**Review by diesieben07 — CHANGES_REQUESTED** [review]

File: patches/minecraft/net/minecraft/world/item/PotionItem.java.patch (hunk 10 above)
Comment (Random832) [reply patches/minecraft/net/minecraft/world/item/PotionItem.java.patch:15]: This is so that potions can also put themselves in tabs other than the brewing tab (same for tipped arrows, re. the combat tab), not only hide themselves from the brewing tab.

**Comment by Shadows-of-Fire** [comment]
I believe all noted concerns have been addressed, the functionality works as expected and the feature is relatively light.  Looks good to me, though I can't resolve the review comments, but that's not critical.
